File: ruby/testcode/unsafereflect_034.rb
require_relative 'shared'

# vuln-code-snippet start ruby_reflect_safe_const_bypass
def handler(req)
  # gsub stripping non-alphanumeric chars is insufficient — namespaced class names with :: pass through
  safe = req.param('klass').gsub(/[^A-Za-z0-9:]/, '')
  klass = Object.const_get(safe) # vuln-code-snippet vuln-line ruby_reflect_safe_const_bypass
  BenchmarkResponse.json({ ok: true, name: klass.name })
end
# vuln-code-snippet end ruby_reflect_safe_const_bypass

File: php/testcode/unsafereflect_034.php
<?php
require_once __DIR__ . '/shared.php';

class PdfHandler {}
class CsvHandler {}

// vuln-code-snippet start php_reflect_allowlist_class_map
function unsafereflect034(BenchmarkRequest $req): BenchmarkResponse {
    $key = $req->param('type');
    $map = ['pdf' => PdfHandler::class, 'csv' => CsvHandler::class]; // vuln-code-snippet safe-line php_reflect_allowlist_class_map
    $cls = $map[$key] ?? null;
    if (!$cls) return BenchmarkResponse::badRequest('invalid');
    $obj = new $cls();
    return BenchmarkResponse::ok('created');
}
// vuln-code-snippet end php_reflect_allowlist_class_map
